## 第 10 回　遺伝子データ解析

### 参考文献：先進ゲノム解析研究推進プラットフォーム (編集)、独習 Python バイオ情報解析、羊土社
---
---
### 目的：ヒト胚性幹細胞を胚様体とした分化過程のシングルセル RNA-Seq 解析データから論文用の図を作成する。

K.R. Moon et al., Visualizing structure and transitions in high-dimensional biological data, Nature Biotechnology, Vol. 37, 1482-1492, 2019.

---
---
## シングルセル解析 1

### テーブルデータの前処理

---
### 細胞の種類

- 細胞には、上皮細胞、神経細胞、筋細胞などたくさんの種類があるが、その構成要素は同じ

<img src="figs/01.jpg" width="1000px">

---
### 細胞
- 細胞は**核**と**細胞質**からなり、物質の出入りを調整する**細胞膜**に包まれている。
- 核には、その細胞の遺伝物質である**染色体**が入っている。

<img src="figs/02.jpg" width="1000px">

---
### 核と染色体

<img src="figs/03.jpg" width="1000px">

---
### 染色体と DNA
<img src="figs/04.jpg" width="1000px">

---
### 受精卵の全能性

<img src="figs/07.jpg" width="1000px">

---
### 受精卵、ヒト胚性幹細胞（Human Embryonic Stem Cells, hESCs）

<img src="figs/09.jpg" width="1000px">

ヒト胚性幹細胞（Human Embryonic Stem Cells, hESCs）は、非常に基本的で重要な細胞の一種です。これらの細胞は、以下の特徴を持っています：

1. **自己複製能力（Self-Renewal）**：ヒト胚性幹細胞は、分化することなく、長期間にわたって自己複製する能力を持っています。つまり、同じ特性を持つ細胞を無限に生成することが可能です。

2. **多能性（Pluripotency）**：ヒト胚性幹細胞は多能性を持っており、これは体内のあらゆる種類の成熟細胞に分化する能力を意味します。つまり、これらの細胞は神経細胞、心筋細胞、皮膚細胞など、人間の体を構成する様々な細胞に変化することができます。

3. **胚からの採取**：これらの細胞は、通常、受精後数日以内の胚から採取されます。この初期段階では、細胞はまだ特定の体の部分や組織に分化していません。

4. **研究と医療への応用**：ヒト胚性幹細胞は、基礎生物学の研究において重要な役割を果たしています。これらの細胞を用いて、細胞の分化過程、疾患のメカニズム、薬剤のスクリーニング、再生医療などが研究されています。

5. **倫理的問題**：ヒト胚性幹細胞の研究は、胚を破壊することになるため、倫理的な議論を引き起こしています。これは、特に胚を潜在的な人間の生命と見なす観点から、多くの議論の対象となっています。

6. **代替技術の開発**：倫理的な懸念に対処するため、科学者たちはiPS細胞（人工多能性幹細胞）などの代替技術を開発しています。これらは成人の体細胞を再プログラムして多能性を持たせる方法で、胚を使用しないため、倫理的な問題が少ないです。

ヒト胚性幹細胞は、そのユニークな特性から再生医療や疾患研究において非常に有望視されていますが、同時に倫理的な問題も伴います。そのため、これらの細胞の使用と研究は、科学的な側面だけでなく、社会的、倫理的な側面も考慮して進められています。

---
### 自己複製能力

<img src="figs/08.jpg" width="1000px">

---
### 胚様体

胚様体（Embryoid body、EB）は、体外で培養された多能性幹細胞（例えば、胚性幹細胞（ES細胞）や誘導多能性幹細胞（iPS細胞））が自発的に形成する3次元の細胞集合体です。胚様体は、幹細胞の分化研究や組織工学、再生医療などの分野で広く利用されています。

#### 胚様体の特徴と重要性

1. **多能性の示現**:
   胚様体は、多能性幹細胞が持つ分化のポテンシャルを示現するモデルとして機能します。これらの細胞集合体は、体内の初期胚発生を模倣し、様々な細胞タイプへの分化能力を持っています。

2. **発生生物学の研究**:
   胚様体は、早期の胚発生過程を研究するための重要なツールです。これにより、細胞分化のメカニズムや発生の初期段階における細胞の振る舞いを理解することができます。

3. **疾患モデルとしての利用**:
   特定の遺伝子を操作した幹細胞から形成される胚様体は、様々な遺伝的疾患のモデルとして使用されます。これにより、疾患のメカニズムの解明や新しい治療法の開発に貢献しています。

4. **薬剤スクリーニング**:
   胚様体は、新しい薬剤の安全性や効果を評価するためのプラットフォームとしても使用されます。多様な細胞タイプへの分化能力を持つため、薬剤の影響を多角的に評価することが可能です。

#### 胚様体の形成

胚様体の形成は、通常、多能性幹細胞を非接着性の条件下で培養することによって行われます。この過程では、細胞間相互作用や微小環境が初期胚発生を模倣する形で変化し、細胞は自発的に集合して球状の構造を形成します。この3次元構造内で、細胞は異なる分化経路をたどり始めます。

#### 注意点

胚様体を用いた研究では、形成される細胞タイプや分化の程度が実験条件に大きく依存するため、培養条件の厳密な管理が必要です。また、胚様体を用いた研究は倫理的な考慮を必要とする場合があります。

胚様体は、幹細胞研究の基本的なツールであり、その応用範囲は医学、生物学、薬学など多岐にわたります。

---
### ES 細胞

<img src="figs/10.jpg" width="1000px">

---
### ES 細胞の問題点

<img src="figs/11.jpg" width="1000px">

---
### iPS 細胞

<img src="figs/12.jpg" width="1000px">

---
### 胚様体分化過程

胚様体（Embryoid Body、EB）の分化過程は、胚性幹細胞（ES細胞）や誘導多能性幹細胞（iPS細胞）が体外で3次元的な構造を形成し、その後、様々な細胞タイプへと分化する過程です。この過程は、自然な胚発生の初期段階を模倣しており、細胞の分化と発生の研究において重要な役割を果たします。

#### 胚様体分化の主要な段階

1. **胚様体の形成**:
   多能性幹細胞は、非接着性の培養条件下で集合し、球状の胚様体を形成します。この過程では、細胞間の相互作用や微小環境が変化し、細胞は分化のためのシグナルを受け取り始めます。

2. **内胚葉、中胚葉、外胚葉の形成**:
   胚様体内で、多能性幹細胞は三つの主要な胚葉（内胚葉、中胚葉、外胚葉）へと分化します。これらの胚葉は、体の様々な組織や器官の起源となります。
   
   - **内胚葉**: 消化管、肝臓、肺などの器官の細胞を形成します。
   - **中胚葉**: 筋肉、骨、血管などの組織を形成します。
   - **外胚葉**: 皮膚、神経系、感覚器官などを形成します。


3. **特定の細胞タイプへの分化**:
   分化した胚葉細胞は、さらに特化した細胞タイプへと分化します。例えば、神経細胞、心筋細胞、血管内皮細胞など、多様な細胞タイプが形成されます。

4. **分化誘導の制御**:
   分化過程は、培養条件、成長因子、化学物質などによって誘導および制御されます。これにより、特定の細胞タイプへの分化を促進または抑制することが可能です。

#### 研究と応用

- **基礎研究**: 胚様体の分化過程は、細胞の分化メカニズムや発生生物学の基本原理を理解するためのモデルとして使用されます。
- **疾患モデル**: 特定の遺伝子を操作した幹細胞から形成される胚様体は、遺伝的疾患のモデルとして利用され、病態の解析や新しい治療法の開発に寄与します。
- **再生医療**: 分化した細胞は、損傷した組織や器官の修復に使用することができます。例えば、心筋細胞は心臓病の治療に、神経細胞は神経変性疾患の治療に応用される可能性があります。

胚様体の分化過程は、細胞の運命決定や組織形成の複雑なプロセスを理解するための貴重なツールですが、その制御や応用にはさらなる研究が必要です。また、倫理的な側面も考慮する必要があります。


---
### シングルセル RNA-Seq 解析（scRNA-Seq）

シングルセルRNAシークエンシング（scRNA-Seq）は、個々の細胞レベルでの遺伝子発現プロファイルを解析する強力な技術です。この手法は、異なる細胞タイプの同定、細胞の発達過程の追跡、疾患状態における細胞の挙動の理解など、多くの生物学的および医学的な疑問に答えるために使用されます。

#### scRNA-Seq解析の基本的なステップ

1. **サンプルの準備**:
   細胞は組織から単離され、個々の細胞がシークエンシングのために準備されます。細胞は生きている状態で、かつできるだけ自然な状態を保つように取り扱われます。

2. **RNAの抽出と逆転写**:
   各細胞からRNAが抽出され、cDNA（補完的DNA）に逆転写されます。このプロセスにより、RNAのシークエンスがDNAの形で保存されます。

3. **ライブラリの準備とシークエンシング**:
   逆転写されたcDNAからシークエンシングライブラリが作成され、次世代シークエンシング（NGS）プラットフォームを使用してシークエンスされます。

4. **データの前処理**:
   シークエンスデータは、品質管理、リードのトリミング、リファレンスゲノムへのマッピングなどの前処理ステップを経て、解析可能な形式に変換されます。

5. **遺伝子発現プロファイルの解析**:
   前処理されたデータから、各細胞の遺伝子発現プロファイルが計算されます。これには、遺伝子ごとのリードカウントや正規化手法が含まれます。

6. **細胞クラスタリングと注釈付け**:
   類似した発現パターンを持つ細胞はクラスタリングアルゴリズムを用いてグループ化され、特定の細胞タイプや状態に注釈が付けられます。

7. **データの解釈と生物学的洞察**:
   クラスタリングされた細胞群の解析を通じて、細胞の種類、発達の経路、疾患状態に関連する遺伝子発現の変化など、生物学的な洞察が得られます。

#### 応用

- **細胞の多様性の理解**: 異なる細胞タイプや細胞のサブタイプを同定し、その機能を理解する。
- **発達生物学**: 細胞の発達過程や分化経路を追跡する。
- **疾患研究**: がん細胞や他の疾患細胞の遺伝子発現の異常を特定し、新たな治療標的を探る。
- **薬剤応答の研究**: 細胞が薬剤にどのように反応するかを理解し、個別化医療への応用を探る。

scRNA-Seqは、細胞レベルでの複雑な生物学的プロセスを解明するための強力なツールですが、データ解析や解釈には専門的な知識が必要です。また、高いコストや技術的な課題もあり、これらの問題を解決するための研究が進行中です。

---
### 遺伝子の発現

<img src="figs/05.jpg" width="1000px">

遺伝子が発現する（遺伝子発現）とは、遺伝子の情報が実際の機能を持つ分子、特にタンパク質やRNA分子に変換される過程を指します。この過程は主に以下のステップで構成されています：

1. **転写 (Transcription)**: 遺伝子発現の最初のステップは、DNAに記録されている遺伝情報をメッセンジャーRNA（mRNA）に転写することです。このプロセスでは、DNAの特定の領域（遺伝子）がRNAポリメラーゼという酵素によってRNAの形式にコピーされます。このmRNAはDNAの遺伝情報の使いやすい形式を提供します。

2. **翻訳 (Translation)**: mRNAは細胞のリボソームに運ばれ、そこでタンパク質へと翻訳されます。リボソームはmRNA上のコドン（3つのヌクレオチドの連続）を読み取り、それに対応するアミノ酸を連結してタンパク質を合成します。

3. **ポストトランスクリプショナルな修飾**: 転写されたmRNAは、細胞の核内外でさまざまな修飾を受けることがあります。これには、キャッピング、ポリアデニル化、スプライシングなどが含まれます。

4. **ポストトランスレーショナルな修飾**: タンパク質が合成された後、それらはしばしばさらなる化学的修飾を受けます。これには、リン酸化、グリコシル化、プロテオリシス（タンパク質の切断）などが含まれます。

遺伝子発現は細胞の種類、発達の段階、外部からの信号などによって異なり、細胞の機能と行動を決定します。遺伝子発現の制御は生物の発達、分化、応答メカニズムなどに不可欠です。

---
---
### データ
- [Moon, et al. 2019 の論文](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3155891)で扱われている、胚様体分化過程のシングルセル RNA-Seq 解析のデータを扱う。
- 27 日間の分化時系列の中で、3 日間の間隔で合計 5 回サンプリングされ、すべてについて [10xGenomics 社](https://www.10xgenomics.com/jp)の装置で cDNA ライブラリを生成してシークエンシング（DNA や RNA を構成する単位であるヌクレオチドの塩基配列を決定すること）を行なっている。
- シークエンス後、[Cell Ranger パイプライン](https://kazumaxneo.hatenablog.com/entry/2020/10/31/120219)によって配列情報から分子バーコード（個別の細胞を区別するためのタグ）ごとの遺伝子のカウント情報まで変換されている。
- フルのデータセットは [Mendeley のデータリポジトリ](https://data.mendeley.com/datasets/v6n743h5ng/1)で公開されている。
  - scRNAseq.zip
- このデータセットから 10% の細胞をランダムにサンプリングした小規模データを使う。

---
### 解析に必要なライブラリのインポート

In [1]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

# 画像出力の設定
sns.set(style='whitegrid')
#matplotlib.rcParams['figure.dpi'] = 200
#matplotlib.rcParams['figure.dpi'] = 150

# それぞれのバージョンの確認
!python --version
print('Numpy:', np.__version__)
print('Pandas:', pd.__version__)
print('Matplotlib:', matplotlib.__version__)
print('Seaborn:', sns.__version__)

Python 3.11.5
Numpy: 1.24.3
Pandas: 2.1.3
Matplotlib: 3.8.1
Seaborn: 0.13.0


---
### フォルダ構造の確認

In [13]:
# tree はディレクトリの階層構造を簡単に確認するためのコマンド

!tree ./scRNAseq

./scRNAseq
├── T0_1A
│   ├── barcodes.tsv
│   ├── desktop.ini
│   ├── genes.tsv
│   └── matrix.mtx
├── T2_3B
│   ├── barcodes.tsv
│   ├── desktop.ini
│   ├── genes.tsv
│   └── matrix.mtx
├── T4_5C
│   ├── barcodes.tsv
│   ├── desktop.ini
│   ├── genes.tsv
│   └── matrix.mtx
├── T6_7D
│   ├── barcodes.tsv
│   ├── desktop.ini
│   ├── genes.tsv
│   └── matrix.mtx
├── T8_9E
│   ├── barcodes.tsv
│   ├── desktop.ini
│   ├── genes.tsv
│   └── matrix.mtx
└── desktop.ini

6 directories, 21 files


T0_1A, T2_3B, T4_5C, T6_7D, T8_9E はそれぞれ、27 日間の分化時系列中で 5 回サンプリングされたそれぞれの時点を表している。  
各時点について、3 つのファイルが格納されている。
- barcodes.tsv
- genes.tsv
- matrix.mtx

---
### barcodes.tsv（バーコードデータ）
**個別の細胞を識別するバーコード**が記述されたテキストファイル

In [1]:
# head はファイルの先頭を表示するコマンド

!head ./scRNAseq/T0_1A/barcodes.tsv

AAACCGTGGCTACA-1
AAACGCTGTAGCGT-1
AAAGATCTGGTACT-1
AAAGATCTTCCTTA-1
AAAGTTTGAGCTCA-1
AACAATACGACAAA-1
AACAATACGCTTCC-1
AACCAGTGGAGGGT-1
AACGGTTGACGCTA-1
AACGTTCTTGACCA-1


---
### genes.tsv（遺伝子データ）
計測された遺伝子が記述されたタブ区切りテキストファイル
- 左側の列：遺伝子の ENSEMBLE Gene ID
- 右側の列：遺伝子のシンボル

In [2]:
# head はファイルの先頭を表示するコマンド

!head ./scRNAseq/T0_1A/genes.tsv

ENSG00000243485	RP11-34P13.3
ENSG00000237613	FAM138A
ENSG00000186092	OR4F5
ENSG00000238009	RP11-34P13.7
ENSG00000239945	RP11-34P13.8
ENSG00000239906	RP11-34P13.14
ENSG00000241599	RP11-34P13.9
ENSG00000279928	FO538757.3
ENSG00000279457	FO538757.2
ENSG00000228463	AP006222.2


---
### matrix.mtx（マトリックスデータ）
各細胞（バーコード）について各遺伝子の発現がいくつ観測されたのかカウント情報をまとめたもの
- Matrix Market (MM) Exchange Formats という形式で記述されている
- 疎行列（含まれる値の多くがゼロであるような行列）を比較的コンパクトに記述するための形式

In [3]:
# head はファイルの先頭を表示するコマンド

!head ./scRNAseq/T0_1A/matrix.mtx

%%MatrixMarket matrix coordinate integer general
%
33694 464 952034
9 18 1
9 43 1
9 61 1
9 102 1
9 347 1
9 358 1
9 363 1


---
### 3 種類のファイルの読込み（T0_1A の場合）

In [14]:
import os
import scipy.io
import numpy as np
import pandas as pd

print('Scipy:', scipy.__version__)

# フォルダ（ディレクトリ）の指定
datadir = './scRNAseq/T0_1A'

# barcodes.tsv の読込み
# pandasでもいいが、このファイルは単純なので Numpy の loadtxt() で読込む
barcodes = np.loadtxt(os.path.join(datadir, 'barcodes.tsv'), dtype='str')

# genes.tsv の読込み
# pandas の read_csv() 関数でタブ区切り（sep='\t'）のテキストを読込む
genes = pd.read_csv(
    os.path.join(datadir, 'genes.tsv'),
    sep='\t',
    index_col=0,
    names=['GeneID', 'Symbol'])

# matrix.mtx の読込み
# 特殊な形式のデータ読み込み。scipy で読込む
mtx = scipy.io.mmread(os.path.join(datadir, 'matrix.mtx')).toarray()

Scipy: 1.11.3


---
### 読込んだデータの確認

#### バーコード（barcodes）: 文字列のリストとして読込んでいる。

In [15]:
# print() 関数で確認

print(barcodes[:10])

['AAACCGTGGCTACA-1' 'AAACGCTGTAGCGT-1' 'AAAGATCTGGTACT-1'
 'AAAGATCTTCCTTA-1' 'AAAGTTTGAGCTCA-1' 'AACAATACGACAAA-1'
 'AACAATACGCTTCC-1' 'AACCAGTGGAGGGT-1' 'AACGGTTGACGCTA-1'
 'AACGTTCTTGACCA-1']


#### 遺伝子（genes）: pandas の DataFrame として、遺伝子 ID とシンボルを対応づけて読込んでいる。

In [16]:
# print()ではなく display()関数を使う。DataFrame を Jupyter で綺麗に表示できる。
display(genes)
#genes # DataFrame を表示するときはこれでも OK

# 保存
genes.to_csv('data/genes.tsv', sep='\t')

Symbol
GeneID                       
ENSG00000243485  RP11-34P13.3
ENSG00000237613       FAM138A
ENSG00000186092         OR4F5
ENSG00000238009  RP11-34P13.7
ENSG00000239945  RP11-34P13.8
...                       ...
ENSG00000277856    AC233755.2
ENSG00000275063    AC233755.1
ENSG00000271254    AC240274.1
ENSG00000277475    AC213203.1
ENSG00000268674       FAM231B

[33694 rows x 1 columns]

#### マトリックス（mtx）: Scipy で読み込んでいる。

In [17]:
# print()関数、display()関数のどちらでも OK

print(mtx)
print()
display(mtx)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]



array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

---
<span style="color: red">**課題 1：**</span> 3 つのデータを dataframe という 1 つの DataFrame に統合してください。ただし、data: mtx、index: GeneID、columns: barcodes とします。そうすると index が遺伝子、columns が細胞の DataFrame を作ることができます。

In [18]:
dataframe = pd.DataFrame(mtx, index=genes.index, columns=barcodes)
display(dataframe)

AAACCGTGGCTACA-1  AAACGCTGTAGCGT-1  AAAGATCTGGTACT-1  \
GeneID                                                                  
ENSG00000243485                 0                 0                 0   
ENSG00000237613                 0                 0                 0   
ENSG00000186092                 0                 0                 0   
ENSG00000238009                 0                 0                 0   
ENSG00000239945                 0                 0                 0   
...                           ...               ...               ...   
ENSG00000277856                 0                 0                 0   
ENSG00000275063                 0                 0                 0   
ENSG00000271254                 0                 0                 0   
ENSG00000277475                 0                 0                 0   
ENSG00000268674                 0                 0                 0   

                 AAAGATCTTCCTTA-1  AAAGTTTGAGCTCA-1  AACAATACGACAAA-1  \
GeneID                                                                  
ENSG00000243485                 0                 0                 0   
ENSG00000237613                 0                 0                 0   
ENSG00000186092                 0                 0                 0   
ENSG00000238009                 0                 0                 0   
ENSG00000239945                 0                 0                 0   
...                           ...               ...               ...   
ENSG00000277856                 0                 0                 0   
ENSG00000275063                 0                 0                 0   
ENSG00000271254                 0                 0                 0   
ENSG00000277475                 0                 0                 0   
ENSG00000268674                 0                 0                 0   

                 AACAATACGCTTCC-1  AACCAGTGGAGGGT-1  AACGGTTGACGCTA-1  \
GeneID                                                                  
ENSG00000243485                 0                 0                 0   
ENSG00000237613                 0                 0                 0   
ENSG00000186092                 0                 0                 0   
ENSG00000238009                 0                 0                 0   
ENSG00000239945                 0                 0                 0   
...                           ...               ...               ...   
ENSG00000277856                 0                 0                 0   
ENSG00000275063                 0                 0                 0   
ENSG00000271254                 0                 0                 0   
ENSG00000277475                 0                 0                 0   
ENSG00000268674                 0                 0                 0   

                 AACGTTCTTGACCA-1  ...  TTGAGGACGTCAAC-1  TTGAGGTGAACGAA-1  \
GeneID                             ...                                       
ENSG00000243485                 0  ...                 0                 0   
ENSG00000237613                 0  ...                 0                 0   
ENSG00000186092                 0  ...                 0                 0   
ENSG00000238009                 0  ...                 0                 0   
ENSG00000239945                 0  ...                 0                 0   
...                           ...  ...               ...               ...   
ENSG00000277856                 0  ...                 0                 0   
ENSG00000275063                 0  ...                 0                 0   
ENSG00000271254                 0  ...                 0                 0   
ENSG00000277475                 0  ...                 0                 0   
ENSG00000268674                 0  ...                 0                 0   

                 TTGCTAACTCGACA-1  TTGCTATGCCTTCG-1  TTGGAGTGAAGGTA-1  \
GeneID                                                                  
ENSG00

---
### 5 個のサンプルの処理
5 個のサンプル（T0_1A, T2_3B, T4_5C, T6_7D, T8_9E）があるので、サンプルの ID だけが異なる同じ処理を for 文で実行し、それぞれの DataFrame を 1 つのリスト dataframes に格納。

In [19]:
# フォルダ（ディレクトリ）の指定
datadir = './scRNAseq'

# DataFrame を格納するためのリスト
dataframes = []

# 5 つのサンプルの処理
for sample_id in ['T0_1A', 'T2_3B', 'T4_5C', 'T6_7D', 'T8_9E']:
    print(sample_id)
    
    # genes.tsv
    genes = pd.read_csv(
        os.path.join(datadir, sample_id, 'genes.tsv'),
        sep='\t',
        index_col=0,
        names=['GeneID', 'Symbol']
    )
    
    # barcodes.tsv
    barcodes = np.loadtxt(
        os.path.join(datadir, sample_id, 'barcodes.tsv'),
        dtype='str'
    )
    
    # matrix.mtx
    mtx = scipy.io.mmread(os.path.join(datadir, sample_id, 'matrix.mtx'))
    mtx = mtx.toarray()
    
    # dataframe
    dataframe = pd.DataFrame(mtx, index=genes.index, columns=barcodes)
    print('\t', dataframe.shape)
    
    # リスト dataframes に dataframe を追加
    dataframes.append(dataframe)

T0_1A
	 (33694, 464)
T2_3B
	 (33694, 737)
T4_5C
	 (33694, 624)
T6_7D
	 (33694, 656)
T8_9E
	 (33694, 632)


---
### 5 個のサンプルの統合
<span style="color: red">**課題 2：**</span> リスト dataframes に 格納された 5 個の dataframe を、pandas の concat を使って、df という名前の 1 つの DataFrame に統合してください。ただし、pandas の keys について調べ、DataFrame の keys をサンプルのラベル（T0_1A, T2_3B, T4_5C, T6_7D, T8_9E）ではなく、'Days0-3', 'Days6-9', 'Days12-15', 'Days18-21', 'Days24-27' という名前を使うこと。

In [21]:
sample_labels = ['Days0-3', 'Days6-9', 'Days12-15', 'Days18-21', 'Days24-27']

df = pd.concat(dataframes, axis=1, keys=sample_labels)

print(f'Total: {len(df)} genes, {len(df.columns)} cells')

df

Total: 33694 genes, 3113 cells


Days0-3                                    \
                AAACCGTGGCTACA-1 AAACGCTGTAGCGT-1 AAAGATCTGGTACT-1   
GeneID                                                               
ENSG00000243485                0                0                0   
ENSG00000237613                0                0                0   
ENSG00000186092                0                0                0   
ENSG00000238009                0                0                0   
ENSG00000239945                0                0                0   
...                          ...              ...              ...   
ENSG00000277856                0                0                0   
ENSG00000275063                0                0                0   
ENSG00000271254                0                0                0   
ENSG00000277475                0                0                0   
ENSG00000268674                0                0                0   

                                                                    \
                AAAGATCTTCCTTA-1 AAAGTTTGAGCTCA-1 AACAATACGACAAA-1   
GeneID                                                               
ENSG00000243485                0                0                0   
ENSG00000237613                0                0                0   
ENSG00000186092                0                0                0   
ENSG00000238009                0                0                0   
ENSG00000239945                0                0                0   
...                          ...              ...              ...   
ENSG00000277856                0                0                0   
ENSG00000275063                0                0                0   
ENSG00000271254                0                0                0   
ENSG00000277475                0                0                0   
ENSG00000268674                0                0                0   

                                                                    \
                AACAATACGCTTCC-1 AACCAGTGGAGGGT-1 AACGGTTGACGCTA-1   
GeneID                                                               
ENSG00000243485                0                0                0   
ENSG00000237613                0                0                0   
ENSG00000186092                0                0                0   
ENSG00000238009                0                0                0   
ENSG00000239945                0                0                0   
...                          ...              ...              ...   
ENSG00000277856                0                0                0   
ENSG00000275063                0                0                0   
ENSG00000271254                0                0                0   
ENSG00000277475                0                0                0   
ENSG00000268674                0                0                0   

                                  ...        Days24-27                   \
                AACGTTCTTGACCA-1  ... TTGCTAACTCGTTT-1 TTGGTACTCGATAC-1   
GeneID                            ...                                     
ENSG00000243485                0  ...                0                0   
ENSG00000237613                0  ...                0                0   
ENSG00000186092                0  ...                0                0   
ENSG00000238009                0  ...                0                0   
ENSG00000239945                0  ...                0                0   
...                          ...  ...              ...              ...   
ENSG00000277856                0  ...                0                0   
ENSG00000275063                0  ...                0                0   
ENSG00000271254                0  ...                0                0   
ENSG00000277475                0  ...                0                0   
ENSG00000268674                0  ...                0                0   

                               

### ここまでの操作をファイルに保存
<span style="color: red">**課題 3：**</span> 巨大な DataFrame である df を、python の **pickle** という仕組みを利用して、**バイナリファイル**として書き出してください。ただし、新たに data というフォルダを作り、その中に df.gz という名前で保存すること。その後、そのアイルを df という名前で読み込んで表示してください。

In [22]:
df.to_pickle('data/df.gz')

---
### クオリティコントロール（細胞と遺伝子のフィルタリング）
エラーやノイズといった異常値を除去する

#### 発現量が 0 の（どの細胞でも発現しなかった）遺伝子を除去
df のインデックスのうち、横方向の和: .sum(axis=1) がゼロであるような遺伝子のインデックスを除去する。各列に細胞が並んでいるので、横の和がゼロ＝いずれの細胞でも発現量ゼロということである。

<span style="color: red">**課題 4：**</span> 発現量が 0 の遺伝子を df から除去してください。

In [23]:
games_zero_values = df.index[df.values.sum(axis=1) == 0]

df.drop(games_zero_values, inplace =True)
print(f'After filterling: {len(df)} genes')

After filterling: 20155 genes


#### 遺伝子が 33694 個あった df が、遺伝子が 20155 個の df になった。 

<span style="color: red">**課題 5：**</span> ここでの df を、python の pickle という仕組みを利用して、バイナリファイルとして data/df_mod.gz として保存してください。

In [24]:
df.to_pickle('data/df.mod.gz')